File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def srachValue(self, row, target):
        n = len(row)
        st = 0
        end = n-1

        while st <= end:
            mid = st + (end -st)//2
            if row[mid] < target:
                st = mid +1
            elif row[mid] > target :
                end = mid -1
            else:
                return True
        return False


    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        row = len(matrix)
        col = len(matrix[0])
        st_row = 0
        end_row = row -1

        while st_row <= end_row:
            mid = st_row + (end_row-st_row)//2
            if matrix[mid][0] <= target <= matrix[mid][col-1]:
                return self.srachValue(matrix[mid], target)
            elif matrix[mid][0] > target:
                end_row = mid -1
            else:
                st_row = mid + 1
        return False

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
#include <vector>
using namespace std;

class Solution {
public:
    bool searchValue(vector<int>& row, int target) {
        int st = 0;
        int end = row.size() - 1;

        while (st <= end) {
            int mid = st + (end - st) / 2;

            if (row[mid] < target)
                st = mid + 1;
            else if (row[mid] > target)
                end = mid - 1;
            else
                return true;
        }
        return false;
    }

    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.empty() || matrix[0].empty())
            return false;

        int rows = matrix.size();
        int cols = matrix[0].size();

        int st_row = 0;
        int end_row = rows - 1;

        while (st_row <= end_row) {
            int mid = st_row + (end_row - st_row) / 2;

            if (matrix[mid][0] <= target && target <= matrix[mid][cols - 1]) {
                return searchValue(matrix[mid], target);
            }
            else if (matrix[mid][0] > target) {
                end_row = mid - 1;
            }
            else {
                st_row = mid + 1;
            }
        }

        return false;
    }
};
